Fix `SymmetricMatrixCoefficient::ProjectSymmetric` bug

j-signorelli opened the pull request:
The matrix defined for projection of SymmetricMatrixCoefficient onto a QuadratureFunction appears to be of incorrect size.

The current code creates a vdim x vdim matrix, where vdim is the dimension of the QF = height*(height+1)/2.

For wanting to output a symmetric 3x3 tensor (6 values), this then creates a 6x6 matrix instead of a 3x3 one. This PR fixes that.
<!--GHEX{"author":"j-signorelli","assignment":"2024-05-11T15:13:42","editor":"tzanio","id":4271,"reviewers":["paul-hilscher","najlkin"],"merge":"2024-05-22T21:03:29","approval":"2024-05-22T04:58:12"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4271](https://github.com/mfem/mfem/pull/4271) | @j-signorelli | @tzanio | @paul-hilscher + @najlkin | 5/11/24 | 5/21/24 | 5/22/24 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

j-signorelli commented:
I am not sure why the linker was not able to find symbols for arm64 for the checks -- if anyone has any suggestions, please let me know!

sebastiangrimberg commented:
> I am not sure why the linker was not able to find symbols for arm64 for the checks -- if anyone has any suggestions, please let me know!

GitHub actions recently started rolling out new runners for  macOS jobs which use macOS 14 and Arm64 instances instead of x64_64. Looks like this is what's causing the problem here -> The cached builds being restored are from an older run on an x86_64 architecture, while the checks for this PR are running on arm64. I could be wrong, but think we need a PR to update the cache keys [here](https://github.com/mfem/mfem/blob/master/.github/workflows/builds-and-tests.yml#L172) which will force rebuilding of the Hypre and METIS dependencies in the cache for PR workflows.

tzanio commented:
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

najlkin reviewed (approved):
Yes, this is a bug for sure, there seems to be confusion between plain `Vector` or `Memory` wrappers and `DenseSymmetricMatrix`, which takes the size of the matrix as an argument of `UseExternalData` and not size of the data.

paul-hilscher commented:
@j-signorelli , 

would it make sense to add a small unit test ? `test_coefficient.cpp` does covers `MatrixCoefficient` but it seems we do not have tests for `SymmetricMatrixCoefficient` / `SymmetricMatrixConstantCoefficient` (ideally including a test for `ProjectSymmetry`).

j-signorelli commented:
@paul-hilscher:
Done, but it required additional changes + bug fixes that probably should be re-reviewed.
- `SymmetricMatrixConstantCoefficient` was broken due to there not being any `DenseSymmetricMatrix &DenseSymmetricMatrix::operator=(const DenseSymmetricMatrix &m)` -- I defined this
- `mat` was being used as an auxiliary variable in `SymmetricMatrixCoefficient` while derived classes `SymmetricMatrixConstantCoefficient` and `SymmetricMatrixFunctionCoefficient` each had their own `mat` variables. I renamed the auxiliary one `mat_aux` for clarification, and moved `GetMatrix()` to only be a member function of `SymmetricMatrixConstantCoefficient`, as it seems to have intended to be based on the documentation for it.
- I removed the override of `Print()` for `DenseSymmetricMatrix` (did this as I was debugging, but it's more convenient to just use `Matrix::Print()` than not have anything implemented anyways)

paul-hilscher commented on linalg/symmat.cpp:
```diff
@@ -58,6 +58,20 @@ DenseSymmetricMatrix &DenseSymmetricMatrix::operator=(real_t c)
+   const int hw = m.GetStoredSize();
+   for (int i = 0; i < hw; i++)
+   {
+      data[i] = m.data[i];
```
`Memory::operator` functions read:
>    /// Array subscript operator for host memory.
   inline T &operator[](int idx);
>   /// Array subscript operator for host memory, const version.
   inline const T &operator[](int idx) const;
   
   will this only be evaluated on host ? Would `Memory::copyTo` be a better option ? (I am not familiar with this piece of code so I am asking to understand this better. Thanks.).

najlkin replied:
It is copied from `DenseMatrix` implementation, where simply we do not care at the moment as anything else than host is not supported 😬

j-signorelli replied:
I am not sure, I just essentially reproduced the code used for `DenseMatrix`'s copy operator (see [here](https://github.com/mfem/mfem/blob/7c296d00d8a770e9f569e4db9d7e6d415902a886/linalg/densemat.cpp#L668-L679)) but with `GetStoredSize()` instead.

paul-hilscher commented on fem/coefficient.hpp:
```diff
@@ -1426,12 +1426,13 @@ class SumCoefficient : public Coefficient
+
    /// Internal matrix used when evaluating this coefficient as a DenseMatrix.
-   DenseSymmetricMatrix mat;
+   mutable DenseSymmetricMatrix mat_aux;
```
minor: maybe `mat_sym`  as it is a bit more descriptive ?  (just a personal preference so feel free to ignore)

najlkin replied:
It is `mat` in child classes vs. `mat_aux` here, where both are symmetric, so it would be rather confusing renaming this one to `mat_sym`. I would keep this, because it indicates it is just a mutable auxiliary thing, which no one should rely on as it changes all the time.

paul-hilscher reviewed (approved):
Thanks for the fix and the tests! Great stuff.

j-signorelli commented on fem/coefficient.hpp:
```diff
@@ -1464,9 +1465,6 @@ class SymmetricMatrixCoefficient : public MatrixCoefficient
                      const IntegrationPoint &ip);
 
-   /// Return a reference to the constant matrix.
-   const DenseSymmetricMatrix& GetMatrix() { return mat; }
```
Done, see [54dcdc7](https://github.com/mfem/mfem/pull/4271/commits/54dcdc720f9bcc7e0c55fc0a895b2bea5740ece6)

j-signorelli commented on tests/unit/fem/test_coefficient.cpp:
```diff
@@ -303,3 +303,36 @@ TEST_CASE("Piecewise Matrix Coefficient", "[Coefficient]")
+   for (int i = 0; i < qfdim; i++)
+   {
+      values[i] = i+1;
+
```
Done, see [9834126](https://github.com/mfem/mfem/pull/4271/commits/98341269cc3c26276592470499fbd2fb3cb24a32)

j-signorelli commented:
@paul-hilscher @najlkin Thank you for your feedback -- if there is anything else to adjust please let me know, otherwise this should be ready to merge into `next` @tzanio

tzanio reviewed (approved):
Thanks @j-signorelli !

tzanio commented:
Merged in `next` for testing...

tzanio commented:
There is a failure with this PR in our regression testing on Lassen (V100 machine) when building optimized + CUDA. The error is below, can you double-check the code?

```
INFO: Test filter: ~[Parallel] 
Filters: ~[Parallel]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
unit_tests is a Catch v2.13.9 host application.
Run with -? for options
-------------------------------------------------------------------------------
Symmetric Matrix Coefficient
-------------------------------------------------------------------------------
/usr/WS2/mfem/gitlab-runner/kolev/builds/zkQzXxmM/001/gitlab/mfem/mfem-autotest/tests/unit/fem/test_coefficient.cpp:307
...............................................................................
/usr/WS2/mfem/gitlab-runner/kolev/builds/zkQzXxmM/001/gitlab/mfem/mfem-autotest/tests/unit/fem/test_coefficient.cpp:332: FAILED:
  REQUIRE( qf.DistanceTo(values) == MFEM_Approx(0.0) )
with expansion:
  0.5017307833147915 == Approx( 0 )
===============================================================================
test cases:     297 |     296 passed | 1 failed
assertions: 2828545 | 2828544 passed | 1 failed
make[1]: *** [makefile:240: unit_tests-test-seq] Error 1
```

najlkin commented:
I think the problem might be that `DistanceTo` sums totally different numbers, it would be better to sum the relative errors, i.e., instead of `(x_i-y_i)^2` sum `(x_i-y_i)^2/y_i^2` :wink:

sebastiangrimberg commented:
> There is a failure with this PR in our regression testing on Lassen (V100 machine) when building optimized + CUDA. The error is below, can you double-check the code?

I'm pretty sure this is because `Vector::DistanceTo` is assumes the vector is not using the device while the added test uses a `QuadratureFunction` which does use the device by default. The following patch to `linalg/vector.hpp` would hopefully fix this:

```
diff --git a/linalg/vector.hpp b/linalg/vector.hpp
index 00bb625f9..292ac6fec 100644
--- a/linalg/vector.hpp
+++ b/linalg/vector.hpp
@@ -678,24 +678,24 @@ inline real_t Distance(const Vector &x, const Vector &y)

 inline real_t Vector::DistanceSquaredTo(const real_t *p) const
 {
-   return DistanceSquared(data, p, size);
+   return DistanceSquared(HostRead(), p, size);
 }

 inline real_t Vector::DistanceSquaredTo(const Vector &p) const
 {
    MFEM_ASSERT(p.Size() == Size(), "Incompatible vector sizes.");
-   return DistanceSquared(data, p.data, size);
+   return DistanceSquared(HostRead(), p.HostRead(), size);
 }

 inline real_t Vector::DistanceTo(const real_t *p) const
 {
-   return Distance(data, p, size);
+   return Distance(HostRead(), p, size);
 }

 inline real_t Vector::DistanceTo(const Vector &p) const
 {
    MFEM_ASSERT(p.Size() == Size(), "Incompatible vector sizes.");
-   return Distance(data, p.data, size);
+   return Distance(HostRead(), p.HostRead(), size);
 }

 /// Returns the inner product of x and y
```

najlkin commented:
Yes, that would explain why it happens only in this context. I presumed these basic methods are safe, but MFEM can always surprise 😄 . It goes quite far beyond the scope of this PR, so it would be perhaps cleaner to just turn off the device processing and open a separate PR for this.
Besides that, I would still recommend the relative errors, but it is not critical and only makes the test slightly insensitive.

sebastiangrimberg commented:
It's a (minor) bug fix so I would encourage adding the patch as part of this PR. But I will let the editor @tzanio decide.

v-dobrev commented:
These "distance" methods were mean for small vectors on CPU where the overhead of `HostRead` (in terms of code and execution time) is not desirable. We can add an MFEM_ASSERT to check they are not used on device. For big vectors that can be on device, one can use subtract and `Normlinf()` or if one really wants $\ell_2$ (i.e. Eucledian) norm, `Norml2()`.

v-dobrev commented on tests/unit/fem/test_coefficient.cpp:
```diff
@@ -303,3 +303,31 @@ TEST_CASE("Piecewise Matrix Coefficient", "[Coefficient]")
+   // Define qf
+   QuadratureFunction qf(qspace, qfdim);
+
+   symCoeff.ProjectSymmetric(qf);
+
+   // Require equality
+   REQUIRE(qf.DistanceTo(values) == MFEM_Approx(0.0));
```
In this case, where `qf` may be on GPU, we can just use `qf.HostRead` before `qf.DistanceTo`.

najlkin replied:
...or subtract `values` and call `Norml2()`, which would be cleaner and better emulate how a quadrature function should be used, but for a small unit test I am fine with both variants :wink:

v-dobrev replied:
It turned out there is no problem with the test -- see https://github.com/mfem/mfem/pull/4271#issuecomment-2118602833.

v-dobrev commented:
Hmm, I'm still not sure how but in my tests with a CUDA build, the unit tests pass. I'm using a debug build but that should not change the result unless it produces an error from a failed check which it does not.

v-dobrev commented:
It turns out the unit test failure is due to a compiler issue: when using `-O3`, IBM XL C++ enables optimizations that can cause issues in some cases and it looks like this is one of those cases. Switching to `-O2` optimization with the same compiler fixes the problem.

v-dobrev commented:
I noticed a potential issue when `SymmetricMatrixCoefficient::ProjectSymmetric` is called with `QuadratureFunction &qf` that is valid on device. Since we are accessing its entries using aliases, and all of that is done on host, it is best to call
```c++
   qf.HostWrite();
```
sometime before the start of the loop over elements, here: https://github.com/mfem/mfem/blob/13e1067cd15326ad2a40251e2fc00fc19bca6f48/fem/coefficient.cpp#L812

tzanio commented:
@j-signorelli what do you think about the discussion and suggestions above ☝️ ?

j-signorelli commented:
@tzanio It sounds like may be it would be best to just include the `qf.HostWrite()` in `ProjectSymmetric`, as @v-dobrev points [out](https://github.com/mfem/mfem/pull/4271#issuecomment-2118622127).

But I have not yet used MFEM on devices, so I defer this entirely up to you and others with more experience.

najlkin commented:
That is ok :wink:, so I did that in a17333cb1918d5550b01b1fdb706c69e450e5436 . I also merged in `master` due to the recent changes of the tests in #4262 .
This should safe the day also with that `DistanceTo`, but we should still fix that (by an assert or `HostRead`), but in a separate PR, I think 🤔 .

najlkin commented:
@tzanio , would you merge in `next` again to see if we are fine or not?

tzanio commented:
Re-merged in `next` for testing...

v-dobrev commented:
I guess no one read my comment regarding the actual source of the issue, so let me re-post it:

> It turns out the unit test failure is due to a compiler issue: when using `-O3`, IBM XL C++ enables optimizations that can cause issues in some cases and it looks like this is one of those cases. Switching to `-O2` optimization with the same compiler fixes the problem.

One solution is to switch to `-O2` with IBM XL C++, another is to switch to GCC. The latter may be a bit simpler to do -- it is just a change in the Spack specs in https://github.com/mfem/mfem/blob/8ed6d6d2d204a282b1c90c165e2d8352581c4eb2/.gitlab/lassen-build-and-test.yml#L17-L25

We'll need to update the pre-built Spack packages used in the nightly tests for this change -- I'll do that and then push the change here.

v-dobrev commented:
Okay, I made the changes necessary to switch to GCC from XL C++. I think the PR can be re-merged in `next`.

v-dobrev commented:
Re-merged in `next` for testing...
